workers/unity/Packages/com.improbable.gdk.transformsynchronization/TransformUtils.cs

```diff
@@ -11,51 +11,146 @@ public static bool HasChanged(Coordinates a, Coordinates b)
+        public static bool HasChanged(FixedPointVector3 a, FixedPointVector3 b)
+        {
+            return a.X != b.X || a.Y != b.Y || a.Z != b.Z;
+        }
+
+        // Decompress from uint32 to Unity Quaternion
+        public static UnityEngine.Quaternion ToUnityQuaternion(this CompressedQuaternion quaternion)
+        {
+            var compressedValue = quaternion.Data;
+            var q = new float[4];
+
+            const uint mask = (1u << 9) - 1u;
+
+            int largestIndex = (int) (compressedValue >> 30);
+            float sumSquares = 0;
+            for (var i = 3; i >= 0; --i)
+            {
+                if (i != largestIndex)
+                {
+                    uint magnitude = compressedValue & mask;
+                    uint signBit = (compressedValue >> 9) & 0x1;
+                    compressedValue >>= 10;
+                    q[i] = SqrtHalf * ((float) magnitude) / mask;
+                    if (signBit == 1)
+                    {
+                        q[i] *= -1;
+                    }
+
+                    sumSquares += Mathf.Pow(q[i], 2);
+                }
+            }
+
+            q[largestIndex] = Mathf.Sqrt(1f - sumSquares);
+
+            return new Quaternion(q[0], q[1], q[2], q[3]);
+        }
+
+        private const float SqrtHalf = 0.70710678118f;
+
+        // Compress from Unity Quaternion to uint32
+        public static CompressedQuaternion ToCompressedQuaternion(this UnityEngine.Quaternion quaternion)
         {
-            return a.W != b.W || a.X != b.X || a.Y != b.Y || a.Z != b.Z;
+            quaternion = quaternion.normalized;
+
+            var q = new float[4] { quaternion.x, quaternion.y, quaternion.z, quaternion.w };
+
+            uint largestIndex = 0;
+            var largestValue = Mathf.Abs(q[largestIndex]);
+            for (uint i = 1; i < 4; ++i)
+            {
+                var componentAbsolute = Mathf.Abs(q[i]);
+                if (componentAbsolute > largestValue)
+                {
+                    largestIndex = i;
+                    largestValue = componentAbsolute;
+                }
+            }
+
+            uint negativeBit = quaternion[(int) largestIndex] < 0 ? 1u : 0u;
+
+            uint compressedQuaternion = largestIndex;
+            for (uint i = 0; i < 4; i++)
+            {
+                if (i != largestIndex)
+                {
+                    uint signBit = (q[i] < 0 ? 1u : 0u) ^ negativeBit;
+                    uint magnitude = (uint) (((1u << 9) - 1u) * (Mathf.Abs(q[i]) / SqrtHalf) + 0.5f);
+                    compressedQuaternion = (compressedQuaternion << 10) | (signBit << 9) | magnitude;
+                }
+            }
+
+            return new CompressedQuaternion(compressedQuaternion);
         }
 
-        public static UnityEngine.Quaternion ToUnityQuaternion(this Quaternion quaternion)
+        public static UnityEngine.Vector3 ToUnityVector3(this FixedPointVector3 fixedPointVector3)
```

Comment:
All of these helper utilities should be normal functions. To avoid users seeing them when using the unity types.